R/utils_render_html.R

```diff
@@ -1081,120 +1039,223 @@ create_footnotes_component_h <- function(data) {
+  alignment_classes <- paste0("gt_", col_alignment)
 
-        # In the above condition, `grand_summary_col` is a global variable
-        # (`"::GRAND_SUMMARY"`) assigned in `dt_summary.R`)
+  for (j in seq_len(nrow(summary_df))) {
 
-        styles_resolved_row <-
-          styles_resolved_group[styles_resolved_group$rownum == j, , drop = FALSE]
+    last_row_class <- "gt_last_summary_row"
 
-      } else {
+    if (summary_row_type == "grand") {
+      styles_resolved_row <- dt_styles_pluck(styles_tbl, locname = "grand_summary_cells", grpname = group_id, rownum = j)
+      summary_row_class <- "gt_grand_summary_row"
+      first_row_class <- "gt_first_grand_summary_row"
 
-        styles_resolved_row <-
-          styles_resolved_group[styles_resolved_group$grprow == j, , drop = FALSE]
-      }
+    } else {
+      styles_resolved_row <- dt_styles_pluck(styles_tbl, locname = "summary_cells", grpname = group_id, grprow = j)
+      summary_row_class <- "gt_summary_row"
+      first_row_class <- "gt_first_summary_row"
+    }
 
-      row_styles <-
-        build_row_styles(
-          styles_resolved_row = styles_resolved_row,
-          stub_available = TRUE,
-          n_cols = n_cols
-        )
+    row_styles <-
+      build_row_styles(
+        styles_resolved_row = styles_resolved_row,
+        include_stub = TRUE,
+        n_cols = n_data_cols
+      )
 
-      summary_row_lines[[length(summary_row_lines) + 1]] <-
-        htmltools::tags$tr(
-          mapply(
-            SIMPLIFY = FALSE,
-            USE.NAMES = FALSE,
-            summary_df_row(j),
-            stub_classes,
-            alignment_classes,
-            row_styles,
-            FUN = function(x, stub_class, alignment_class, cell_style) {
-
-              if (j == 1) {
-                summary_row_class <- c(summary_row_class, first_row_class)
-              }
+    summary_row_lines[[length(summary_row_lines) + 1]] <-
+      htmltools::tags$tr(
+        htmltools::HTML(
+          paste0(
+            mapply(
+              SIMPLIFY = FALSE,
+              USE.NAMES = FALSE,
+              unname(unlist(summary_df[j, ])),
+              col_span_vals,
+              alignment_classes,
+              extra_classes,
+              row_styles,
+              FUN = function(x, col_span, alignment_class, extra_class, cell_style) {
 
-              htmltools::tags$td(
-                class = paste(
-                  c("gt_row", stub_class, alignment_class, summary_row_class),
-                  collapse = " "
-                ),
-                style = cell_style,
-                htmltools::HTML(x)
-              )
-            }
+                extra_class <- c(extra_class, summary_row_class)
+
+                if (j == 1) {
+                  extra_class <- c(extra_class, first_row_class)
+                }
+
+                if (j == nrow(summary_df)) {
+                  extra_class <- c(extra_class, last_row_class)
+                }
+
+                sprintf(
```

Comment:
Any user-provided values that are being inserted into the HTML using sprintf/paste, need to be escaped using htmltools::htmlEscape(..., attribute = TRUE).